### PR TITLE
Expose tiller server metrics to prometheus

### LINKED ISSUE #2163: Expose tiller metrics to prometheus
We would like to expose various tiller metrics to Prometheus using https://github.com/grpc-ecosystem/go-grpc-prometheus . I am thinking about exposing this via `/metrics` path in probe port.

What do you think?

### PR DESCRIPTION
**Pull request by sadlil**
Fixes #2163

### CODE REVIEW

**Comment by thomastaylor312** [comment]
I am going to hold off on putting this in a milestone until more discussion can happen on this.

**Comment by technosophos** [comment]
@fibonacci1729 and @jchauncey -- Could I beg you two to pair on this and validate that this is the correct approach for collecting Prometheus metrics? Admittedly, this is outside my expertise.

**Comment by jchauncey** [comment]
Here is a sample of the metrics that [go-grpc-prometheus](https://github.com/grpc-ecosystem/go-grpc-prometheus) exposes

```
# HELP go_gc_duration_seconds A summary of the GC invocation durations.
# TYPE go_gc_duration_seconds summary
go_gc_duration_seconds{quantile="0"} 3.4359e-05
go_gc_duration_seconds{quantile="0.25"} 4.5177e-05
go_gc_duration_seconds{quantile="0.5"} 5.5868e-05
go_gc_duration_seconds{quantile="0.75"} 8.0119e-05
go_gc_duration_seconds{quantile="1"} 0.000207363
go_gc_duration_seconds_sum 0.001685209
go_gc_duration_seconds_count 23
# HELP go_goroutines Number of goroutines that currently exist.
# TYPE go_goroutines gauge
go_goroutines 17
# HELP go_memstats_alloc_bytes Number of bytes allocated and still in use.
# TYPE go_memstats_alloc_bytes gauge
go_memstats_alloc_bytes 4.147952e+06
# HELP go_memstats_alloc_bytes_total Total number of bytes allocated, even if freed.
# TYPE go_memstats_alloc_bytes_total counter
go_memstats_alloc_bytes_total 7.9673296e+07
# HELP go_memstats_buck_hash_sys_bytes Number of bytes used by the profiling bucket hash table.
# TYPE go_memstats_buck_hash_sys_bytes gauge
go_memstats_buck_hash_sys_bytes 1.487611e+06
# HELP go_memstats_frees_total Total number of frees.
# TYPE go_memstats_frees_total counter
go_memstats_frees_total 665234
# HELP go_memstats_gc_sys_bytes Number of bytes used for garbage collection system metadata.
# TYPE go_memstats_gc_sys_bytes gauge
go_memstats_gc_sys_bytes 739328
# HELP go_memstats_heap_alloc_bytes Number of heap bytes allocated and still in use.
# TYPE go_memstats_heap_alloc_bytes gauge
go_memstats_heap_alloc_bytes 4.147952e+06
# HELP go_memstats_heap_idle_bytes Number of heap bytes waiting to be used.
# TYPE go_memstats_heap_idle_bytes gauge
go_memstats_heap_idle_bytes 7.340032e+06
# HELP go_memstats_heap_inuse_bytes Number of heap bytes that are in use.
# TYPE go_memstats_heap_inuse_bytes gauge
go_memstats_heap_inuse_bytes 6.651904e+06
# HELP go_memstats_heap_objects Number of allocated objects.
# TYPE go_memstats_heap_objects gauge
go_memstats_heap_objects 22738
# HELP go_memstats_heap_released_bytes_total Total number of heap bytes released to OS.
# TYPE go_memstats_heap_released_bytes_total counter
go_memstats_heap_released_bytes_total 6.955008e+06
# HELP go_memstats_heap_sys_bytes Number of heap bytes obtained from system.
# TYPE go_memstats_heap_sys_bytes gauge
go_memstats_heap_sys_bytes 1.3991936e+07
# HELP go_memstats_last_gc_time_seconds Number of seconds since 1970 of last garbage collection.
# TYPE go_memstats_last_gc_time_seconds gauge
go_memstats_last_gc_time_seconds 1.490370241239965e+09
# HELP go_memstats_lookups_total Total number of pointer lookups.
# TYPE go_memstats_lookups_total counter
go_memstats_lookups_total 302
# HELP go_memstats_mallocs_total Total number of mallocs.
# TYPE go_memstats_mallocs_total counter
go_memstats_mallocs_total 687972
# HELP go_memstats_mcache_inuse_bytes Number of bytes in use by mcache structures.
# TYPE go_memstats_mcache_inuse_bytes gauge
go_memstats_mcache_inuse_bytes 9600
# HELP go_memstats_mcache_sys_bytes Number of bytes used for mcache structures obtained from system.
# TYPE go_memstats_mcache_sys_bytes gauge
go_memstats_mcache_sys_bytes 16384
# HELP go_memstats_mspan_inuse_bytes Number of bytes in use by mspan structures.
# TYPE go_memstats_mspan_inuse_bytes gauge
go_memstats_mspan_inuse_bytes 107920
# HELP go_memstats_mspan_sys_bytes Number of bytes used for mspan structures obtained from system.
# TYPE go_memstats_mspan_sys_bytes gauge
go_memstats_mspan_sys_bytes 196608
# HELP go_memstats_next_gc_bytes Number of heap bytes when next garbage collection will take place.
# TYPE go_memstats_next_gc_bytes gauge
go_memstats_next_gc_bytes 8.15904e+06
# HELP go_memstats_other_sys_bytes Number of bytes used for other system allocations.
# TYPE go_memstats_other_sys_bytes gauge
go_memstats_other_sys_bytes 1.834493e+06
# HELP go_memstats_stack_inuse_bytes Number of bytes in use by the stack allocator.
# TYPE go_memstats_stack_inuse_bytes gauge
go_memstats_stack_inuse_bytes 688128
# HELP go_memstats_stack_sys_bytes Number of bytes obtained from system for stack allocator.
# TYPE go_memstats_stack_sys_bytes gauge
go_memstats_stack_sys_bytes 688128
# HELP go_memstats_sys_bytes Number of bytes obtained by system. Sum of all system allocations.
# TYPE go_memstats_sys_bytes gauge
go_memstats_sys_bytes 1.8954488e+07
# HELP grpc_server_handled_total Total number of RPCs completed on the server, regardless of success or failure.
# TYPE grpc_server_handled_total counter
grpc_server_handled_total{grpc_code="OK",grpc_method="UninstallRelease",grpc_service="hapi.services.tiller.ReleaseService",grpc_type="unary"} 1
# HELP grpc_server_msg_received_total Total number of RPC stream messages received on the server.
# TYPE grpc_server_msg_received_total counter
grpc_server_msg_received_total{grpc_method="UninstallRelease",grpc_service="hapi.services.tiller.ReleaseService",grpc_type="unary"} 1
# HELP grpc_server_msg_sent_total Total number of gRPC stream messages sent by the server.
# TYPE grpc_server_msg_sent_total counter
grpc_server_msg_sent_total{grpc_method="UninstallRelease",grpc_service="hapi.services.tiller.ReleaseService",grpc_type="unary"} 1
# HELP grpc_server_started_total Total number of RPCs started on the server.
# TYPE grpc_server_started_total counter
grpc_server_started_total{grpc_method="UninstallRelease",grpc_service="hapi.services.tiller.ReleaseService",grpc_type="unary"} 1
```

**Comment by jchauncey** [comment]
@technosophos I think this is fine for now. Its definitely a good start to monitoring the tiller process and prometheus is use extensively in other parts of the kubernetes infrastructure. I think it would be nice to have another issue that starts a conversation on the types of metrics tiller should expose that are not covered by this lib.

**Comment by thomastaylor312** [comment]
@technosophos Are we ok with putting this in 2.3?

**Comment by technosophos** [comment]
@thomastaylor312  No, we will put this on 2.4 and get it merged as soon as tests are passing.

**Comment by sadlil** [comment]
@technosophos @thomastaylor312 All tests are passing.

**Comment by technosophos** [comment]
A quick rebase, and we can get this tested and merged into 2.4.0

File: cmd/tiller/tiller.go [hunk 1]
```diff
@@ -130,6 +131,11 @@ func start(c *cobra.Command, args []string) {
 
 	go func() {
 		mux := newProbesMux()
+
```
Comment (technosophos) [inline cmd/tiller/tiller.go]: Is there any way to disable prometheus collection?

**Comment by sadlil** [comment]
@technosophos rebased, tests passing.

>  Is there any way to disable prometheus collection?
Currently No. But there is scope to add flag to tiler that can disable prometheus collection.

**Review by technosophos — APPROVED** [review]
Manually tested on system without prometheus enabled in order to test that case. And all seemed to work fine.

**Pull request merged**